Ble extended advertising

pan- opened the pull request:
### Description

This PR introduce extended and periodic advertising management in mbed-os BLE API. These two features have been introduced in Bluetooth 5 and greatly improve advertisement capabilities of compatible devices. 

#### Extended Advertising 

With extended advertising, devices are not limited to a single payload advertised and a single scan response. Devices can manage multiple concurrent advertisement data called advertising sets as they contain their own data and operate according to their own parameters. 

 Other noticeable improvement of extended advertising are: 
* Advertising on Coded PHY for long range advertisement on the main channel. All phys are available to the secondary channel. 
* Increased payload size: from 31 byte to 251 in a single packet and up to 1560 byte when fragmentation is used. 
* Control of TX power.

#### Periodic Advertising 

Periodic advertising is a new kind of advertising that allows a device to stream data at a fixed rate. Peers can sync on the data stream and listen to it. Internally it is very similar to a unidirectional connection. 

#### Integration 

On many aspects it wasn't possible to add this new feature on the top of the previous GAP advertisement API. Defects such as (incorrect) global address type, global filter policies, state bookkeeping and global scanning/advertisement parameters mutable in a non atomic way would have make it memory heavy and not very usable as it would have been full of corner cases (it already is!) due to the global state held in it if we wanted to preserve backward compatibility.  

The new API introduced is basically a version two of the GAP Advertisement/Scanning/Connection API. It is **not** compatible with the old subsystem. You have to use one or the other but it is forbidden to use both concurrently. That way we preserve complete backward compatibility and prepare a clean ground to build upcoming Bluetooth features.

#### TODO

- [x] Review
- [x] Add deprecated notices where relevant
- [x] Improve documentation
- [x] Move deprecated code around (nice to have)


### Pull request type

<!-- 
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [ ] Fix
    [ ] Refactor
    [ ] Target update
    [X] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

paul-szczepanek-arm committed:
fix names

paul-szczepanek-arm committed:
ppm conversion

paul-szczepanek-arm committed:
connection complete event

paul-szczepanek-arm committed:
fix swapping

paul-szczepanek-arm committed:
fix first element of array selection

paul-szczepanek-arm committed:
rename phy to index

paul-szczepanek-arm committed:
missing return

paul-szczepanek-arm committed:
assert added

paul-szczepanek-arm committed:
removed impossible case

paul-szczepanek-arm committed:
we need clamping while we wait for c++17

paul-szczepanek-arm committed:
clamp values

paul-szczepanek-arm committed:
set random address

paul-szczepanek-arm committed:
cancel connection

paul-szczepanek-arm committed:
only one connection event needed

paul-szczepanek-arm committed:
ppm case

paul-szczepanek-arm committed:
default values and func names

paul-szczepanek-arm committed:
event types fixed

paul-szczepanek-arm committed:
fix review comments

paul-szczepanek-arm committed:
units

pan- committed:
Merge pull request #1 from paul-szczepanek-arm/vincent

connections

paul-szczepanek-arm committed:
new units

paul-szczepanek-arm committed:
comparison operator

paul-szczepanek-arm committed:
new units

pan- committed:
BLE: Fix mutability in ble::advertising_event_t

paul-szczepanek-arm committed:
docs

pan- committed:
BLE: Move common type from pal to ble namespace.

pan- committed:
BLE: Add phy in sync exstablished event.

pan- committed:
BLE: Rework Gap event handler.

Every user facing event handler accept a single parameter type; event that contains all the event fields.
That strategy allows us to extend the type later if required while we do not disrupt existing code.

paul-szczepanek-arm committed:
disallow removing the legacy handle

paul-szczepanek-arm committed:
more docs

paul-szczepanek-arm committed:
merge fail

paul-szczepanek-arm committed:
merge cleanup

paul-szczepanek-arm committed:
docs

paul-szczepanek-arm committed:
docs

paul-szczepanek-arm committed:
docs

paul-szczepanek-arm committed:
docs

pan- committed:
BLE: Introduce Duration and Bounded classes.

paul-szczepanek-arm committed:
docs and synchronise connection result with other events

paul-szczepanek-arm committed:
docs

pan- committed:
Update mbed-os/features/FEATURE_BLE/ble/Gap.h

Co-Authored-By: paul-szczepanek-arm <paul.szczepanek@arm.com>

pan- committed:
Update mbed-os/features/FEATURE_BLE/ble/Gap.h

Co-Authored-By: paul-szczepanek-arm <paul.szczepanek@arm.com>

paul-szczepanek-arm committed:
docs

paul-szczepanek-arm committed:
docs

paul-szczepanek-arm committed:
docs

pan- committed:
BLE: Use secure types in GapScanParameters.

pan- committed:
BLE: Address special case when advertising payload size is 0.

pan- committed:
BLE: Deal with random address rotation where appropriate.

pan- committed:
BLE: Initialisation of legacy advertising set.

pan- committed:
BLE: Add PHY precondition check in setScanParameters.

pan- committed:
BLE: Add strict Duration type

* Construction from representation type is explicit
* Conversion between unit is only allowed if there is no precision loss
* Range is built inside the type
* More operators!
* min and max values exposed as part of the interface

pan- committed:
BLE: Move GapScanParameters header into ble/gap/ScanParameters

pan- committed:
BLE: rename GapExtendedConnectParameters_t into ConnectionParameters

pan- committed:
BLE: Expose ble::pal::connection_role_t in ble namespace.

pan- committed:
BLE: Move Gap event types in dedicated file.

paul-szczepanek-arm committed:
docs

paul-szczepanek-arm committed:
docs

pan- committed:
BLE: Extract types from GapAdvertisingConstant and add them into ble namespace.

donatieng commented on features/FEATURE_BLE/ble/BLETypes.h:
```diff
@@ -32,6 +33,237 @@
+
+typedef Duration<uint32_t, 1> microsecond_t;
+typedef Duration<uint32_t, 1000> millisecond_t;
+typedef Duration<uint32_t, 1000> second_t;
```
Three zeros are missing :)

pan- replied:
fixed.

paul-szczepanek-arm commented on features/FEATURE_BLE/source/generic/GenericGap.cpp:
```diff
@@ -1555,5 +1641,970 @@ void GenericGap::set_connection_event_handler(pal::ConnectionEventMonitor::Event
+    }
+
+    if (is_extended_advertising_available()) {
+        return _pal_gap.extended_scan_enable(
```
we allow the user to set scanning parameters that are then ignored

for example the user may set coded phy scanning but because it is not supported this will run a normal scan but return success

we can't reject it in the settings either because we use chaining in the sets that are incapable of returning error codes

this needs to check the settings and report an error if any of the extended scan params have been set

pan- replied:
Parameters are rejected on legacy controllers if they include phy parameters: 

https://github.com/ARMmbed/mbed-os/blob/3a2fb280f07770d6f5fa17f0465c716203665d22/features/FEATURE_BLE/source/generic/GenericGap.cpp#L2383-L2385

However we need to check scan and connection parameter; please do not resolve the conversation.

pan- committed:
BLE: Move ble::Duration in its own file.

paul-szczepanek-arm commented on features/FEATURE_BLE/ble/BLETypes.h:
```diff
@@ -32,6 +33,237 @@
+
+
+template<typename Rep, uint32_t TB, typename Range = typename DefaultRange<Rep>::type >
+struct Duration {
```
put this in its own header, this isn't ble specific and would be useful to other people too!

pan- replied:
Done.

Would be nice to have it in platform some day but it requires move refinement.

pan- committed:
move ble::Bounded in its own file.

pan- committed:
BLE: rename duration type and extract them out of Gap.

pan- committed:
BLE: move GapAdvertisingParameters to ble::AdvertisingParameters

donatieng commented on features/FEATURE_BLE/ble/gap/AdvertisingDataBuilder.h:
```diff
@@ -0,0 +1,552 @@
+        _payloadLen++;
+
+        /* Field ID. */
+        memset(_buffer.data() + _payloadLen, (uint8_t)advDataType, 1);
```
Offset issue?

pan- replied:
Not really, the payload length is incremented by one one line above.

pan- committed:
BLE: Move ConnectionParameters into ble namespace.

donatieng commented on features/FEATURE_BLE/ble/gap/AdvertisingDataBuilder.h:
```diff
@@ -0,0 +1,552 @@
+        }
+
+        /* Field length. */
+        memset(_buffer.data() + _payloadLen, len + 1, 1);
```
Why use `memset` here?

pan- replied:
No particular reason. Fixed.

donatieng commented on features/FEATURE_BLE/ble/gap/AdvertisingParameters.h:
```diff
@@ -0,0 +1,699 @@
+     * @return True if channel used.
+     */
+    bool getChannel38() const {
+        return _channel37;
```
37 -> 38

pan- replied:
fixed.

donatieng commented on features/FEATURE_BLE/ble/gap/AdvertisingParameters.h:
```diff
@@ -0,0 +1,699 @@
+     * @return True if channel used.
+     */
+    bool getChannel39() const {
+        return _channel37;
```
37 -> 39

pan- replied:
fixed.

paul-szczepanek-arm committed:
move feature support into user api

donatieng commented on features/FEATURE_BLE/ble/gap/ConnectionParameters.h:
```diff
@@ -0,0 +1,300 @@
+    }
+
+    /** Handle the swapping of 2M and CODED so that the array is ready for the pal call. */
+    void swapCodedAnd2M() {
```
Seems a bit convoluted? What's the rationale behind the way these arrays are handled? And why we have multiple arrays of variables instead of a single array of structs?

paul-szczepanek-arm replied:
because the pal call requires the parameters to be arranged in memory like that - arrays of parameters for each phy, and they need to be contiguous even if there are only 2 of them - hence the swapping

pan- replied:
@donatieng I'm resolving this. I think everyone agree it ain't pretty but it is an implementation detail. We can rework it latter if required; it will just cost 48 bytes of stack.

pan- committed:
BLE: move ScanParameters into ble namespace.

donatieng reviewed (dismissed):
Haven't looked at the PAL yet but the API looks great. really good work there. Some notes/questions.

paul-szczepanek-arm committed:
fix case

pan- committed:
BLE: cleanup ble::AdvertisingParameters

* Use constrained types where relevant.
* Remove fields N/A fields

pan- committed:
BLE: Use common ble::own_address_type_t in ble::ScanParameters.

pan- committed:
BLE: unwrap own_address_type value in GenericGap.

pan- committed:
BLE: Fix second time base.

pan- committed:
BLE: Fix AdvertisingParameters::getChannelXX functions.

0xc0170 commented:
@pan- All comments addressed? And ready for another round of review (helps reviewers if you add a comment after rebase/adding more commits - "whats new").

pan- committed:
BLE: Improve ble::AdvertisingParameters field declaration.

pan- committed:
BLE: Rework ble::AdvertisingParameters

* Reorganize functions in logical order.
* Add separation between legacy and extended advertising.
* Update setter to return a reference to this. This allows chain of calls.
* Remove getter that can fail due to bad user inputs.

pan- committed:
BLE: Fix missing initialization in ble::phy_set_t.

pan- committed:
BLE: Fix ternary expression return type.

pan- committed:
BLE: Improve readability of AdvertisingDataBuilder::appendField

pan- committed:
BLE: Move new gap types in ble/gap/Types.h

paul-szczepanek-arm committed:
reject conenct request if exteneded not enabled but required

pan- committed:
BLE: Move leftover target_peer_address_type_t in ble/gap/Types.h

pan- committed:
BLE: merge ble::advertising_type_t and pal::advertising_type_t.

pan- committed:
BLE: Merve ble::advertising_policy_mode and pal::advertising_filter_policy .

pan- committed:
BLE: Revert space change in ble/GapAdvertisingParams.h

pan- committed:
BLE: merge ble::scanning_policy_mode_t and gap::scanning_filter_policy_t

This change set also introduce ble::initiator_filter_policy_t and fix a type issue ble::ConnectionParameters.

pan- committed:
BLE: merge ble::scanning_filter_duplicates_t and pal::duplicate_filter_t

paul-szczepanek-arm committed:
update gap overview

pan- committed:
BLE: merge ble::own_address_t and pal::own_address_type_t.

paul-szczepanek-arm committed:
advertising data

paul-szczepanek-arm committed:
size clarification

paul-szczepanek-arm committed:
fix names

pan- committed:
BLE: merge ble::target_peer_address_type_t and advertising_peer_address_type_t

paul-szczepanek-arm committed:
move values out of GAP

pan- committed:
BLE: cleanup of clock_accuracy_t

pan- committed:
BLE: Cleanup ble::connection_role_t.

pan- committed:
BLE: remove not required prefixes in ble::ConnectionParameters.

pan- committed:
BLE: Move gap events into ble namespace.

pan- committed:
BLE: remove ble:: prefixes in ble::ScanParameters.

pan- committed:
BLE: remove ble:: prefixes in ble::AdvertisingParameters.

cmonr commented:
@donatieng Mind re-reviewing? Looks like your comments were addressed.

cmonr commented:
And @paul-szczepanek-arm, although it seems like you're involved in the PR itself 😄

pan- committed:
BLE: rename Gap::IsFeatureSupported into Gap::isFeatureSupported

pan- committed:
BLE: rename functions that trap concurrent use of incorect GAP APIs

pan- committed:
BLE: remove inner V2 typedef in Gap class.

pan- committed:
BLE: Add explicit constructor for Duration that accepts millisecond in input.

pan- committed:
BLE: clean separation between GAP v1 and GAP v2 API.

pan- commented:
@cmonr  That's a collective effort 😉 .

pan- committed:
BLE: Introduce new disconnection API.

paul-szczepanek-arm committed:
extra data builder funcs

paul-szczepanek-arm committed:
remove data instead of replace data

pan- commented on features/FEATURE_BLE/ble/gap/AdvertisingDataBuilder.h:
```diff
@@ -516,56 +515,296 @@ class AdvertisingDataBuilder
+    ) {
+        uint8_t* field = findField(advDataType);
+
+        if (field) {
```
Would it be possible to have a clear semantic where: 
* `set`: That add a new field if it doesn't exist or update it if it exist.
* `clear`: remove an existing field.

paul-szczepanek-arm replied:
> * `set`: That add a new field if it doesn't exist or update it if it exist.

I have split setField into two: addField and replaceField as I already know which operation I'm performing

> * `clear`: remove an existing field.

there is a removeField which I think is less ambiguous becuase clear may mean "clear contents" but retain field, remove field is non-ambiguous

pan- commented on features/FEATURE_BLE/ble/gap/AdvertisingDataBuilder.h:
```diff
@@ -516,56 +515,296 @@ class AdvertisingDataBuilder
+     * COMPLETE_LIST_128BIT_SERVICE_IDS or LIST_128BIT_SOLICITATION_IDS, the
+     * supplied value is will not be appended but replaced completely.
+     */
+    ble_error_t removeData(
```
Why is this name removeData if it add or update a field ?

pan- commented on features/FEATURE_BLE/ble/gap/AdvertisingDataBuilder.h:
```diff
@@ -516,56 +515,296 @@ class AdvertisingDataBuilder
+     * adv_data_type_t::APPEARANCE as the field type.
+     */
+    ble_error_t setAppearance(
+        adv_data_appearance_t appearance = adv_data_appearance_t::GENERIC_TAG
```
please remove the default parameter.

pan- commented on features/FEATURE_BLE/ble/gap/AdvertisingDataBuilder.h:
```diff
@@ -516,56 +515,296 @@ class AdvertisingDataBuilder
+        * adv_data_type_t::FLAGS as the field type.
+        */
+       ble_error_t setFlags(
+           adv_data_flags_t flags = adv_data_flags_t::LE_GENERAL_DISCOVERABLE
```
Default flag should be `BREDR_NOT_SUPPORTED | LE_GENERAL_DISCOVERABLE`. 
`adv_data_flags_t` needs to be updated as it is not an enum but a set of flags.

pan- commented on features/FEATURE_BLE/ble/gap/AdvertisingDataBuilder.h:
```diff
@@ -516,56 +515,296 @@ class AdvertisingDataBuilder
+    * element being the length of the field followed by the value of the field.
+    * @return NULL if the field is not present in the payload.
+    */
+    uint8_t* findField(adv_data_type_t type)
```
Is it possible to return a span to the field value ?

paul-szczepanek-arm replied:
yes but I'm going to immediately cast to pointer and ignore the size so I don't see the point of pretending like it's a span

pan- replied:
Why would you ignore the size of the field ? that is the whole point of returning a Span.

paul-szczepanek-arm replied:
because I have to operate on the whole payload, not just the span of the field, I have to move data between boundaries of spans

pan- commented on features/FEATURE_BLE/ble/gap/AdvertisingDataBuilder.h:
```diff
@@ -516,56 +515,296 @@ class AdvertisingDataBuilder
+        if (field) {
+            switch(advDataType.value()) {
+                /* These types are append to existing field */
+                case adv_data_type_t::INCOMPLETE_LIST_16BIT_SERVICE_IDS:
```
It should be possible to replace these types even if it requires a memmove.

paul-szczepanek-arm replied:
you can replace them by calling removeData(adv_data_type_t::INCOMPLETE_LIST_16BIT_SERVICE_IDS) first

pan- replied:
That is not consistent with the other data type.

pan- reviewed (commented):
AdvDataBuilder review: 

It would be good if we add new helpers functions: 

```c++
setFlags(adv_flags_t);
setName(const char*, bool complete);
setTxPowerLevel(int8_t);
setAppearance(adv_appearance_t);
setManufacturerSpecificData(Span<const uint8_t>);
setLocalServiceList(Span<const UUID>, bool complete);
setConnectionIntervalPreference(conn_interval_t min, conn_interval_t max);
setServiceData(UUID serviceUUID, Span<const uint8_t> serviceData);
setAdvertisingInterval(adv_interval_t); // note: serialized to a uint16_t!
setRequestedServices(Span<const UUID>); 
```

paul-szczepanek-arm committed:
flags no longer a safenum

paul-szczepanek-arm committed:
fix enum values

paul-szczepanek-arm committed:
helper functinos

pan- commented on features/FEATURE_BLE/ble/gap/AdvertisingDataBuilder.h:
```diff
@@ -516,56 +515,296 @@ class AdvertisingDataBuilder
+                    return appendToField(fieldData, field);
+                default:
+                    /* All other types have their field contents replaced */
+                    return replaceField(advDataType, fieldData, field);
```
If add means add can we return an error if the field is already present ?

paul-szczepanek-arm replied:
addField doesn't check if the field is already there.

1. check field exists with findField()
2a). if field exists replaceField() or appendToField() for special fields
2b) if field doesn't exist addField()

In my implementation the private field functions are low level blocks that addData and removeData use.

paul-szczepanek-arm replied:
Usage of add and remove in addData and removeData is different but that's the public interface.

pan- committed:
BLE: Add API to manage GAP connection parameter updates.

paul-szczepanek-arm committed:
UUID handling

paul-szczepanek-arm committed:
remove redundant calulations

paul-szczepanek-arm committed:
service data

paul-szczepanek-arm committed:
individual operations

paul-szczepanek-arm commented:
what's new

AdvertisingDataBuilder updated with helper functions (but not complete, will post when done)

paul-szczepanek-arm committed:
CONNECTION INTERVAL HELPER

paul-szczepanek-arm committed:
docs

NirSonnenschein commented:
@pan- : this needs a rebase

paul-szczepanek-arm committed:
verify parameters

paul-szczepanek-arm committed:
requested services

paul-szczepanek-arm committed:
check sizes, use memcpy

paul-szczepanek-arm committed:
simplify UUID funcs

paul-szczepanek-arm committed:
check min size for custom data

paul-szczepanek-arm committed:
fix size checks

paul-szczepanek-arm committed:
allow long uuids in service data

donatieng commented on features/FEATURE_BLE/ble/Gap.h:
```diff
@@ -82,44 +87,84 @@ class GapAdvertisingData;
+ * extended advertising however even if the controller supports larger data lengths if
+ * you wish to be compatible with older devices you may wish to use legacy
+ * advertising and should not use payloads larger than LEGACY_ADVERTISING_MAX_SIZE.
  *
```
Might be worth mentioning that this feature was introduced in Bluetooth 5 and that 5 support from the controller's standpoint is necessary but not sufficient.

pan- replied:
@donatieng I thin I addressed your comment.

pan- committed:
BLE: Move new version of ::GAP in a new class ::ble::Gap .

donatieng commented on features/FEATURE_BLE/ble/Gap.h:
```diff
@@ -82,44 +87,84 @@ class GapAdvertisingData;
+ * Subsequently you may disable extended advertising and the periodic advertising
+ * will continue. If you start periodic advertising while extended advertising is
+ * inactive, periodic advertising will not start until you start extended advertising
+ * at a later time.
```
Might be worth adding a quick word on why that's required (background from spec)

pan- replied:
@donatieng I thin I addressed your comment.

donatieng commented on features/FEATURE_BLE/ble/Gap.h:
```diff
@@ -321,6 +269,617 @@ class GapAdvertisingData;
+     */
+    virtual bool isFeatureSupported(ble::controller_supported_features_t feature)
+    {
+        // TODO: deal with legacy implementation
```
I guess this should return `false` by default

pan- replied:
Fixed

donatieng commented on features/FEATURE_BLE/ble/Gap.h:
```diff
@@ -321,6 +269,617 @@ class GapAdvertisingData;
+         *
+         * @param event Scan request event: @see ScanRequestEvent_t for details.
+         */
+        virtual void onScanRequest(const ble::ScanRequestEvent_t& event) { }
```
Would this method (and similar) need to be renamed to something like `onScanRequestReceived` to be consistent with other names like `onPeriodicAdvertisingSyncEstablished`?

pan- replied:
@donatieng Fixed.

donatieng commented on features/FEATURE_BLE/ble/Gap.h:
```diff
@@ -321,6 +269,617 @@ class GapAdvertisingData;
+    virtual ble_error_t setAdvertisingPayload(
+        ble::advertising_handle_t handle,
+        mbed::Span<uint8_t> payload,
+        bool minimiseFragmentation = false
```
Might be worth explaining why you'd ever set `minimiseFragmentation` to `false` (why would I prefer fragmentation ;)?). Also it should be `minimizeFragmentation` as we're supposed to write in US-English :)

pan- replied:
I fixed it with a bazooka; I removed the parameter as it is not very useful for the user right now.
It may be reintroduced latter if it is necessary.

paul-szczepanek-arm committed:
service data allows both types of UUIDs

donatieng commented on features/FEATURE_BLE/ble/Gap.h:
```diff
@@ -321,6 +269,617 @@ class GapAdvertisingData;
+     *
+     * @return BLE_ERROR_NONE on success.
+     */
+    virtual ble_error_t startScan(
```
I'm not sure how easy that would be to add to the `Duration` class, but could we have a special `Duration::forever` value?

pan- replied:
That is something we can add when it is meaningful: 

```c++
scan_duration_t::forever(); // works 
scan_interval_t::forever(); // compile time error as there is no special forever value for scan interval.
```

@donatieng What do you think ?

donatieng replied:
Sounds good :)

paul-szczepanek-arm committed:
use correct type for service data based on uuid length

donatieng reviewed (commented):
A few more notes, thanks for addressing my comments so far

pan- committed:
BLE: prefix inclusion of Gap header with "ble/"

paul-szczepanek-arm committed:
pull in inherited overloads

bulislaw reviewed (approved):
The "not mixing" new and old APIs sounds reasonable to me. Someone else needs to review the code.

pan- committed:
BLE: Move Gap deprecated implementation in source/Gap.cpp .

The implementation has been moved to a block that do not issue deprecation warning as this is an implementation detail.

pan- committed:
BLE: Move uses of deprecated Gap functions into isolated and warning free block.

pan- committed:
BLE: provide default implementation ::ble::Gap::isFeatureSupported

pan- committed:
BLE: prevent warning from deprecated API implementation in GenericGap.

paul-szczepanek-arm committed:
solve link problems

pan- committed:
BLE: Fix signeness warning in AdvertisingDataBuilder.

paul-szczepanek-arm committed:
added extra constructor

paul-szczepanek-arm committed:
fixed index and default constructor in conn params

paul-szczepanek-arm commented:
gap/ConnectionParameters.h fixed (as in needs re-review)

paul-szczepanek-arm committed:
missing defaults

pan- committed:
BLE: Move Gap::StopScan into ::ble::Gap::stopScan .

pan- committed:
BLE: Move privacy subsystem to ::ble::Gap.

pan- committed:
BLE: Fix return value of GenericGap::manageConnectionParametersUpdateRequest.

pan- committed:
BLE: Fix includes of ble/gap/AdvertisingDataBuilder.h ,

pan- committed:
BLE: Split ble/gap/AdvertisingDataBuilder.h into declaration and implementation file.

pan- committed:
BLE: return immutable buffer from AdvertisingDataBuilder::getAdvertisingData

paul-szczepanek-arm committed:
fix missing index

paul-szczepanek-arm committed:
pull in more inherited overloads

paul-szczepanek-arm committed:
missing virtual

paul-szczepanek-arm committed:
missing overloads

paul-szczepanek-arm committed:
fix constness to match recent const changes

pan- committed:
BLE: Cleanup bounded and document it.

pan- committed:
BLE: Add Duration::forever() function.

pan- committed:
BLE: Fix Bounded constant declaration.

pan- committed:
BLE: Handle forever value in ble::Duration operators.

pan- committed:
BLE: Improve static constant of Range and Duration.

paul-szczepanek-arm committed:
missing virtual

paul-szczepanek-arm committed:
added type conversion for connection params

pan- committed:
BLE: Document ble::Duration and satelite classes.

paul-szczepanek-arm committed:
moved some funcs to cpp, moved cpp to source folder

pan- committed:
BLE: Add forever information to ble::adv_duration_t and ble::scan_duration_t

pan- committed:
BLE: Fix constant declaration of ble::AdvertisingParameters.

cmonr commented:
Me thinks this needs a rebase.

paul-szczepanek-arm committed:
return func made const

pan- committed:
BLE: rename AdvertisingParameters::*PolicyMode into AdvertisingParameters::*Filter.

pan- committed:
BLE: Improve ble::AdvertisingParameters documentation.

paul-szczepanek-arm committed:
duration added valueInMs

paul-szczepanek-arm committed:
simple scan paramater constructor

0xc0170 commented:
@ARMmbed/mbed-os-pan When is this completed ? 

It still needs work and review

0xc0170 commented:
While we are reviewing, I'll run this through CI

pan- committed:
BLE: Document ble::ConnectionParameters.

pan- committed:
BLE: set the phy as first parameter in ble::ConnectionParameters.

pan- committed:
BLE: Add code example for ble::ConnectionParameters.

pan- committed:
BLE: Improve naming consistency of ble::Gap::EventHandler functions.

adbridge commented on features/FEATURE_BLE/ble/BLETypes.h:
```diff
@@ -32,6 +33,53 @@
 
 namespace ble {
 
+/* replace with std::clamp when it arrives */
```
Is this the right kind of comment we should have here? Would something like 'preliminary clamp implementation' be better ?

pan- replied:
Remove. It is not used anyway.

adbridge commented on features/FEATURE_BLE/ble/BLETypes.h:
```diff
@@ -578,6 +638,13 @@ struct peer_address_type_t :SafeEnum<peer_address_type_t, uint8_t> {
+         *
+         * @note This value can be used to indicate the absence of phy
+         */
+        NONE = 0,
```
Should this have an LE_ prefix to be consistent with all the other constants in this enum?

pan- replied:
I would keep it as `NONE` as it is a special value. Other members are prefixed with `LE_` because we can't start an identifier with a number (`1M`, `2M`).

What do you think ?

adbridge replied:
That's fine

adbridge commented on features/FEATURE_BLE/ble/Gap.h:
```diff
@@ -3161,4 +2818,55 @@ class Gap {
+ble_error_t Gap::startScan(
+    T *object,
+    void (T::*callbackMember)(const AdvertisementCallbackParams_t *params)
+) {
```
'{' should be on a newline and not the same line as the function declaration

pan- replied:
fixed.

adbridge commented on features/FEATURE_BLE/ble/common/Bounded.h:
```diff
@@ -0,0 +1,107 @@
+     *
+     * @return The current value.
+     */
+    Rep value() const {
```
'{' should be on a  newline of it's own

pan- replied:
fixed.

adbridge commented on features/FEATURE_BLE/ble/common/Duration.h:
```diff
@@ -0,0 +1,506 @@
+     *
+     * @return The duration in TB units.
+     */
+    Rep value() const {
```
'{' on its own line

pan- replied:
fixed.

adbridge commented on features/FEATURE_BLE/ble/common/Duration.h:
```diff
@@ -0,0 +1,506 @@
+     *
+     * @return The duration in milliseconds.
+     */
+    uint32_t valueInMs() const {
```
'{' on its own line

pan- replied:
fixed.

adbridge commented on features/FEATURE_BLE/ble/common/Duration.h:
```diff
@@ -0,0 +1,506 @@
+     * Return the duration value meaning forever.
+     * @return the duration value meaning forever.
+     */
+    static Duration forever()
```
So there is a structure called 'Forever' and a function called 'forever' this could be very confusing and the only distinction being a capital 'F' doesn't seem very good.

pan- replied:
Fixed. I renamed `Forever` into `Value` as it is a compile time value.

adbridge commented on features/FEATURE_BLE/ble/common/Duration.h:
```diff
@@ -0,0 +1,506 @@
+ * @return The converted duration. It is rounded up if precision is loss.
+ */
+template<typename DurationOut, typename RepIn, uint32_t TBIn, typename RangeIn, typename FIn>
+DurationOut durationCast(Duration<RepIn, TBIn, RangeIn, FIn> duration) {
```
'{' on its own line

pan- replied:
fixed.

adbridge commented on features/FEATURE_BLE/ble/common/Duration.h:
```diff
@@ -0,0 +1,506 @@
+Duration<Rep, TB, Range, F> operator+(
+    Duration<Rep, TB, Range, F> lhs,
+    Duration<Rep, TB, Range, F> rhs
+) {
```
'{' on its own line

pan- replied:
fixed.

adbridge commented on features/FEATURE_BLE/ble/common/Duration.h:
```diff
@@ -0,0 +1,506 @@
+ * @return A duration that represents the multiplication of lhs with rhs.
+ */
+template<typename Rep, uint32_t TB, typename Range, typename F>
+Duration<Rep, TB, Range, F> operator*(Duration<Rep, TB, Range, F> lhs, uint32_t rhs) {
```
'{' on its own line

pan- replied:
fixed.

adbridge commented on features/FEATURE_BLE/ble/common/Duration.h:
```diff
@@ -0,0 +1,506 @@
+ * @return A duration that represents the multiplication of lhs with rhs.
+ */
+template<typename Rep, uint32_t TB, typename Range, typename F>
+Duration<Rep, TB, Range, F> operator*(uint32_t lhs, Duration<Rep, TB, Range, F> rhs) {
```
'{' on its own line

pan- replied:
fixed.

adbridge commented on features/FEATURE_BLE/ble/common/Duration.h:
```diff
@@ -0,0 +1,506 @@
+    typename RepLHS, uint32_t TBLHS, typename RangeLHS, typename FLHS,
+    typename RepRHS, uint32_t TBRHS, typename RangeRHS, typename FRHS
+>
+bool operator<(Duration<RepLHS, TBLHS, RangeLHS, FLHS> lhs, Duration<RepRHS, TBRHS, RangeRHS, FRHS> rhs) {
```
'{' on its own line

pan- replied:
fixed.

adbridge commented on features/FEATURE_BLE/ble/common/Duration.h:
```diff
@@ -0,0 +1,506 @@
+ * @return true if lhs is less than rhs and false otherwise.
+ */
+template<typename Rep, uint32_t Us, typename Range, typename F>
+bool operator<(Duration<Rep, Us, Range, F> lhs, Duration<Rep, Us, Range, F> rhs) {
```
'{' on its own line

pan- replied:
fixed.

adbridge commented on features/FEATURE_BLE/ble/common/Duration.h:
```diff
@@ -0,0 +1,506 @@
+bool operator<=(
+    Duration<RepLHS, TBLHS, RangeLHS, FLHS> lhs,
+    Duration<RepRHS, TBRHS, RangeRHS, FRHS> rhs
+) {
```
'{' on its own line

pan- replied:
fixed.

adbridge commented on features/FEATURE_BLE/ble/common/Duration.h:
```diff
@@ -0,0 +1,506 @@
+ * @return true if lhs is less than or equal to rhs and false otherwise.
+ */
+template<typename Rep, uint32_t Us, typename Range>
+bool operator<=(Duration<Rep, Us, Range> lhs, Duration<Rep, Us, Range> rhs) {
```
'{' on its own line

pan- replied:
fixed.

adbridge commented on features/FEATURE_BLE/ble/common/Duration.h:
```diff
@@ -0,0 +1,506 @@
+bool operator==(
+    Duration<RepLHS, TBLHS, RangeLHS, FLHS> lhs,
+    Duration<RepRHS, TBRHS, RangeRHS, FRHS> rhs
+) {
```
'{' on its own line

pan- replied:
fixed.

adbridge commented on features/FEATURE_BLE/ble/common/Duration.h:
```diff
@@ -0,0 +1,506 @@
+ * @return true if lhs is equal to rhs and false otherwise.
+ */
+template<typename Rep, uint32_t Us, typename Range, typename F>
+bool operator==(Duration<Rep, Us, Range, F> lhs, Duration<Rep, Us, Range, F> rhs) {
```
I'm finding lots of instances where the opening '{' for a function definition is on the same line as the function declaration. It should be on the following newline. Please check all instances in this file.

pan- replied:
fixed.

adbridge commented on features/FEATURE_BLE/ble/gap/AdvertisingDataBuilder.h:
```diff
@@ -0,0 +1,929 @@
+    /** Create from raw value */
+    adv_data_flags_t(uint8_t value = 0) : _value(value) {};
+
+    adv_data_flags_t& setGeneralDiscoverable(bool enable = true) {
```
'{' on its own line

pan- replied:
fixed.

adbridge commented on features/FEATURE_BLE/ble/gap/AdvertisingDataBuilder.h:
```diff
@@ -0,0 +1,929 @@
+        return *this;
+    }
+
+    adv_data_flags_t& setLimitedDiscoverable(bool enable = true) {
```
Same comment as per previous file, please check *all* new files for conformance to opening '{' for function definitions being on a newline.

pan- replied:
fixed.

adbridge commented on features/FEATURE_BLE/ble/gap/ConnectionParameters.h:
```diff
@@ -0,0 +1,465 @@
+        }
+
+        if (is_swapped && phy == phy_t::LE_CODED) {
+            index -= 1;
```
Could this go out of bounds ?

paul-szczepanek-arm replied:
no, phy_t::LE_CODED is last in the array and there must be at least 3 in the array for this mechanic to be required

adbridge commented on features/FEATURE_BLE/ble/gap/ConnectionParameters.h:
```diff
@@ -0,0 +1,465 @@
+    {
+        uint8_t index;
+        switch(phy.value()) {
+            case phy_t::LE_1M:
```
So an explicit NONE type was added to the phy_t but doesn't seem to be used?

paul-szczepanek-arm replied:
will be handled by default case

adbridge replied:
Yes I understand that, my question was more to do with , why bother to add the NONE type in the first place if it isn't used ?

paul-szczepanek-arm replied:
NONE is not a valid phy therefore there is no index for it in this array which only stores possible phys.

(BTW the reason for the the elaborate swapping and indexing is because the underlying layer requires individual arrays of parameter so they need to be contiguous in memory)

pan- committed:
BLE: Update documentation of Gap::Events.

pan- committed:
BLE: Exclude Gap Event constructors from documentation.

adbridge commented on features/FEATURE_BLE/ble/gap/Gap.h:
```diff
@@ -0,0 +1,803 @@
+         *
+         * @param event Scan request event: @see ScanRequestEvent_t for details.
+         */
+        virtual void onScanRequestReceived(const ScanRequestEvent_t &event) { }
```
I believe according to the coding guidelines that empty functions should be:
function_xxxx
{
}

pan- replied:
fixed.

adbridge commented on features/FEATURE_BLE/ble/gap/Types.h:
```diff
@@ -0,0 +1,664 @@
+     */
+    bool connectable() const
+    {
+        return static_cast<bool>(value & (1 << 0));
```
Shift by 0 ? Why ?
Also subsequent shifts below should be explained or enum'd.

pan- replied:
Fixed.

adbridge commented on features/FEATURE_BLE/ble/gap/Types.h:
```diff
@@ -0,0 +1,664 @@
+     */
+    advertising_data_status_t data_status() const
+    {
+        return static_cast<advertising_data_status_t::type>((value >> 5) & 0x03);
```
Shift and mask should be explained or enum'd

pan- replied:
Fixed. Let me know if the fix suits you.

adbridge commented on features/FEATURE_BLE/ble/gap/Types.h:
```diff
@@ -0,0 +1,664 @@
+     */
+    uint16_t get_ppm() {
+        switch(value()) {
+            case PPM_500: return 500;
```
Why not just actual define PPM_xxx as xxx then this function would be unecessary ? Seems a very weird thing to do.

pan- replied:
The value of the enum is the value mandated by the standard.

adbridge commented on features/FEATURE_BLE/ble/generic/GenericGap.h:
```diff
@@ -47,10 +47,13 @@ class GenericGap : public ::Gap,
                    public pal::Gap::EventHandler {
 
 public:
+    /* TODO: move to config */
```
Shouldn't have 'To Do's in released code!

pan- replied:
Serious question, why ? We're documenting a future improvement.

adbridge replied:
@bulislaw What's your feelings on ToDos in released code ?

pan- replied:
Note that the `TODO` is in a private header!

adbridge commented on features/FEATURE_BLE/ble/pal/GapTypes.h:
```diff
@@ -518,19 +332,332 @@ struct received_advertising_type_t :
+struct extended_advertising_report_event_type_t {
+    extended_advertising_report_event_type_t(uint8_t value) : value(value) { }
+
+    bool connectable() {
```
Again shifts and masks with no explanation!

pan- replied:
I removed the type as it lives in gap/Types.h (it is the other one you flagged with the not defined constants.)

adbridge commented on features/FEATURE_BLE/source/Gap.cpp:
```diff
@@ -70,17 +70,22 @@ Gap::PeerAddressType_t convert_legacy_address_type(
+} // end of anonymous namespace
+
+uint8_t Gap::getMaxWhitelistSize(void) const
+{
```
Bracket consistency! General comment coming at the end of the review

pan- replied:
fixed.

adbridge commented on features/FEATURE_BLE/source/Gap.cpp:
```diff
@@ -113,32 +118,175 @@ void Gap::processConnectionEvent(
+)  {
+    /* Avoid compiler warnings about unused variables. */
+    (void)valueArrayPP;
+    (void)countP;
```
Why is this parameter being voided and then used below ?

pan- replied:
Cut and paste from the old code. I will fix that.

pan- replied:
Fixed.

adbridge commented on features/FEATURE_BLE/source/gap/AdvertisingDataBuilder.cpp:
```diff
@@ -0,0 +1,469 @@
+) {
+    uint8_t* field = findField(advDataType);
+
+    if (field) {
```
Presumably this is doing a boolean test of the return value of a pointer , assuming that NULL pointer is false and any other pointer value is true. Bit hacky mixing of types. Pointer values should really be explicitly tested against NULL.

pan- replied:
I look at the standard, it is true that pre-c++11 `NULL` value is implementation defined; it is not the case in C++11 where NULL must be 0. 
At the same time all C++ standards mandate that NULL evaluate to 0. 
So, the behaviour is perfectly defined in that context. 

I can add the NULL evaluation.

adbridge replied:
It is more for clarity of types. Booleans should be treated as such and pointers should be treated as pointers. Even testing for a 0 value using a boolean operation isn't very nice, although perfectly legal

pan- replied:
Fixed.

adbridge commented on features/FEATURE_BLE/source/gap/AdvertisingDataBuilder.cpp:
```diff
@@ -0,0 +1,469 @@
+    adv_data_appearance_t appearance
+) {
+    uint8_t appearence_byte = appearance.value();
+    mbed::Span<const uint8_t> appearance_span((const uint8_t*) &appearence_byte, 2);
```
What does the 2 mean ? Probably should be commented or #defined 
Similarly with values below.

pan- replied:
It was incorrect to begin with. Fixed.

pan- committed:
BLE: Improve naming consistency of Gap event types.

adbridge commented on features/FEATURE_BLE/source/gap/AdvertisingDataBuilder.cpp:
```diff
@@ -0,0 +1,469 @@
+ble_error_t AdvertisingDataBuilder::setManufacturerSpecificData(
+    mbed::Span<const uint8_t> data
+) {
+    if (data.size() < 2) {
```
Why < 2 ? What does this mean ? Should be commented or #defined

pan- replied:
Fixed.

adbridge commented on features/FEATURE_BLE/source/gap/AdvertisingDataBuilder.cpp:
```diff
@@ -0,0 +1,469 @@
+
+    return addOrReplaceData(
+        adv_data_type_t::ADVERTISING_INTERVAL,
+        mbed::make_Span((const uint8_t*)interval.storage(), 2)
```
Again magic number

pan- replied:
Fixed.

adbridge commented on features/FEATURE_BLE/source/gap/AdvertisingDataBuilder.cpp:
```diff
@@ -0,0 +1,469 @@
+    conn_interval_t max
+) {
+    uint32_t interval =  max.value();
+    interval = interval << 16;
```
Why shift by 16 ? Comment or define

pan- replied:
Fixed.

adbridge commented on features/FEATURE_BLE/source/gap/AdvertisingDataBuilder.cpp:
```diff
@@ -0,0 +1,469 @@
+    interval |= min.value();
+    return addOrReplaceData(
+        adv_data_type_t::SLAVE_CONNECTION_INTERVAL_RANGE,
+        mbed::make_Span((const uint8_t*)&interval, 4)
```
Magic number 4

pan- replied:
Fixed.

adbridge commented on features/FEATURE_BLE/source/gap/AdvertisingDataBuilder.cpp:
```diff
@@ -0,0 +1,469 @@
+    adv_data_type_t short_type = adv_data_type_t::SERVICE_DATA_16BIT_ID;
+    adv_data_type_t long_type = adv_data_type_t::SERVICE_DATA_128BIT_ID;
+
+    size_t total_size = data.size() + service.getLen() + 2;
```
Magic number 2

pan- replied:
Fixed.

adbridge commented on features/FEATURE_BLE/source/gap/AdvertisingDataBuilder.cpp:
```diff
@@ -0,0 +1,469 @@
+{
+    /* Scan through advertisement data */
+    for (uint8_t idx = 0; idx < _payload_length; ) {
+        uint8_t fieldType = _buffer[idx + 1];
```
Why the +1 on the index

pan- replied:
Added a documentation bloc explaining advertising data format.

adbridge commented on features/FEATURE_BLE/source/gap/AdvertisingDataBuilder.cpp:
```diff
@@ -0,0 +1,469 @@
+    adv_data_type_t advDataType,
+    mbed::Span<const uint8_t> fieldData
+) {
+    if (fieldData.size() > 0xFE) {
```
Why 0xFE ? Define ?

pan- replied:
Fixed.

adbridge commented on features/FEATURE_BLE/source/gap/AdvertisingDataBuilder.cpp:
```diff
@@ -0,0 +1,469 @@
+    }
+
+    /* Make sure we don't exceed the buffer size */
+    if (_payload_length + fieldData.size() + 2 > _buffer.size()) {
```
Why size + 2 ?

pan- replied:
Fixed.

adbridge commented on features/FEATURE_BLE/source/gap/AdvertisingDataBuilder.cpp:
```diff
@@ -0,0 +1,469 @@
+        }
+    }
+
+    if ((size_long * 8 > 0xFE) || (size_short * 2 > 0xFE)) {
```
How is anyone going to know what this means? So many magic numbers...

pan- replied:
@paul-szczepanek-arm Could you solve that one ?

paul-szczepanek-arm replied:
fixed

adbridge commented on features/FEATURE_BLE/source/gap/AdvertisingDataBuilder.cpp:
```diff
@@ -0,0 +1,469 @@
+    }
+
+    /* calculate total size including headers for types */
+    size_t total_size = size_long + (!!size_long) * 2 +
```
An explanation of exactly why this calculation is structured this way would be most helpful

pan- replied:
@paul-szczepanek-arm Could you solve that one ?

paul-szczepanek-arm replied:
fixed

adbridge commented on features/FEATURE_BLE/source/gap/Gap.cpp:
```diff
@@ -0,0 +1,341 @@
+uint8_t Gap::getMaxAdvertisingSetNumber()
+{
+    /* Requesting action from porter(s): override this API if this capability is supported. */
+    return 1;
```
Magic number

pan- replied:
Fixed.

adbridge commented on features/FEATURE_BLE/source/gap/Gap.cpp:
```diff
@@ -0,0 +1,341 @@
+uint8_t Gap::getMaxAdvertisingDataLength()
+{
+    /* Requesting action from porter(s): override this API if this capability is supported. */
+    return 0x1F;
```
Magic number

pan- replied:
Fixed.

adbridge commented on features/FEATURE_BLE/source/generic/GenericGap.cpp:
```diff
@@ -589,6 +615,107 @@ ble_error_t GenericGap::connect(
+
+    if (!is_extended_advertising_available()) {
+        phy_set_t set(connectionParams.getPhySet());
+        if (set.count() != 1 || set.get_1m() == false) {
```
Why '== false' instead of just '!' ?

pan- replied:
It seemed more readable at the time. Is it an issue ?

cmonr replied:
It might be if we had a recommended consistent way of addressing this.

This time, it's not.

adbridge commented on features/FEATURE_BLE/source/generic/GenericGap.cpp:
```diff
@@ -589,6 +615,107 @@ ble_error_t GenericGap::connect(
+    conn_event_length_t minConnectionEventLength,
+    conn_event_length_t maxConnectionEventLength
+) {
+    if (supervisionTimeout <= (1 + slaveLatency.value()) * maxConnectionInterval * 2) {
```
Magic numbers

pan- replied:
Fixed.

adbridge commented on features/FEATURE_BLE/source/generic/GenericGap.cpp:
```diff
@@ -589,6 +615,107 @@ ble_error_t GenericGap::connect(
+    conn_event_length_t minConnectionEventLength,
+    conn_event_length_t maxConnectionEventLength
+) {
+    if (supervisionTimeout <= (1 + slaveLatency.value()) * maxConnectionInterval * 2) {
```
Magic numbers

pan- replied:
Fixed.

adbridge commented on features/FEATURE_BLE/targets/TARGET_CORDIO/source/CordioPalGap.cpp:
```diff
@@ -398,32 +399,197 @@ void Gap::gap_handler(const wsfMsgHdr_t* msg) {
+                (ble::phy_t::type)evt->txPhy,
+                (ble::phy_t::type)evt->rxPhy
+            );
+        } break;
```
Unless it is the style already adopted throughout this file, then pretty sure 'break' should be on it's own line and shouldn't it be inside the { } ? Though again it should be checked against the coding standards as to whether { } should be being used with case statements anyway.

pan- replied:
How would you declare your local case variables without opening a scope ? 
I can place break statement one line bellow `}`.

pan- replied:
I fixed style of this file too.

adbridge reviewed (changes requested):
General comments:

1. All new files MUST conform to the coding guidelines (camel case variables are an agreed exception to this due to the legacy coding)
2. For changes to existing files, the styling MUST match what is already there.

Throughout this PR there are inconsistencies with the use of { and } for function definitions. Coding guidelines state this format:

function_xxxx
{
}

AFAIK this is also the case for empty functions. I have highlighted some instances where this is wrong but by no means all.

There are also many instances where magic numbers are used with no explanation whatsoever. This leads to a real maintenance nightmare in the future. All magic numbers should have their use either commented or be #defined . Again I have highlighted some instances but by no means all. In some cases the numbers are actually used in mathematical formulae , indexing etc which make no sense unless you know the code inside out. 

There are some real coding queries too, so please check each comment.

adbridge commented:
@mbed-os-pan please address review comments

pan- committed:
BLE: Improve ble::Gap::EventHandler documentation.

pan- committed:
BLE: Fix function braces opening in header files.

pan- committed:
BLE: Fix code formating issues in Gap subsystem.

pan- committed:
BLE: Remove ble::clamp as it is not used anywhere.

pan- committed:
BLE: rename ble::Forever into ble::Value.

pan- committed:
BLE: Fix compiler warning about unused variable.

pan- committed:
BLE: remove ble::pal::extended_advertising_report_event_type_t as it is not used.

pan- committed:
BLE: Improve implementation readability of ble::advertising_event_t

pan- committed:
BLE: Address code formating issue in CordioPalGap implementation.

pan- committed:
BLE: Explain magic constant in ble::Gap implementation.

pan- committed:
BLE: expose representation type and pointer to the inner storage in SafeEnum.

pan- committed:
BLE: Expose representation type in ble::Duration.

pan- committed:
BLE: Improve code by explaining or textually defining constants

pan- committed:
BLE: Improve NULL check in ble::AdvertisingDataBuilder implementation.

pan- committed:
BLE: Extract computation of minimumSupervisionTimeout.

pan- committed:
BLE: remove minimize fragmentation parameter in Gap.

0xc0170 commented:
CI started (review shall still continue), just to get some tests after the last one here (2 days ago)

pan- committed:
BLE: Update ble::Gap documentation.

pan- committed:
BLE: fix name of functions in ble::ScanParameters.

pan- committed:
BLE: rework ScanParameters::phy_configuration_t

pan- committed:
BLE: Construct a phy_set_t from a phy_t.

pan- committed:
BLE: Rework ScanParameters construction.

pan- committed:
BLE: Document ble::ScanParameters.

pan- committed:
BLE: Reorganize ble::gap documentation group.

NirSonnenschein commented:
CI re-triggered due to additional commits after previous run.

NirSonnenschein commented:
@pan- : note this now needs a rebase before it can be merged

pan- committed:
BLE: Improve gap/Types.h documentation.

pan- committed:
BLE: Cleanup enum duplicates in gap/Types.h

pan- committed:
BLE: Improve Duration documentation.

pan- committed:
BLE: move privacy configuration type in ble namespace.

pan- committed:
BLE: doxygen fixes.

NirSonnenschein commented:
@pan- If you want to run CI on the new changes today, please let me know (and also rebase to fix the conflict).

pan- committed:
BLE: Move advertising data types in their own file.

pan- committed:
BLE: Add an AdvertisingDataParser

pan- committed:
BLE: Add a simple builder for AdvertisingData.

pan- commented:
@NirSonnenschein Thanks, I will ask for a run tomorrow morning as a rebase may impact others working on the code if we do not sync first.

paul-szczepanek-arm commented on features/FEATURE_BLE/ble/gap/AdvertisingDataParser.h:
```diff
@@ -0,0 +1,122 @@
+    /**
+     * Return the next advertising data element.
+     *
+     * @note Calling this function if there is no next element is undefined
```
it needn't be, an empty element can be returned

pan- replied:
Yes it does, if you haven't called `hasNext` before `next` and there is no elements left it is like dereferencing an array without checking the bound.

paul-szczepanek-arm committed:
fix size calculation for UUID data payload

paul-szczepanek-arm committed:
fix wrong variable being swapped

paul-szczepanek-arm committed:
replace consts with a private enum

paul-szczepanek-arm committed:
simplify and explain swapping

paul-szczepanek-arm committed:
replace magic numbers

paul-szczepanek-arm committed:
replace magic numbers

paul-szczepanek-arm committed:
make PHY case consistent

pan- committed:
BLE: Improve ble::Gap documentation.

pan- commented:
@AnotherButler I think this one is good for docs review.

pan- committed:
BLE: Fix address type in ::ble::Gap::connect

pan- committed:
BLE: Fallback to legacy connect when required.

bulislaw commented:
All the PRs need to be engineering ready (marked as "needs: CI") by the end of the day (Austin time). Otherwise it won't make 5.11 and will need to come in the next release (5.12 for features, 5.11.1 for fixes and new platforms).

adbridge reviewed (approved):
Review updates look good AFAICT

pan- committed:
BLE: return phy_set_t in ConnectionParameters::getPhySet

pan- committed:
BLE: initialize properly deprecated and non deprecated flags.

pan- committed:
BLE: Mark LEGACY_ADVERTISING_HANDLE as active in initialization.

pan- committed:
BLE: Check supervision timeout value in connect.

pan- committed:
BLE: Implement connection parameter management in GAP.

pan- committed:
BLE: tag API uses in GenericGap.

pan- commented:
@donatieng @paul-szczepanek-arm Could you review today's changes ? 
I think we're good on this one, the only piece missing being the loading of extended API in the Cordio Stack if it is supported by the controller. 

I will update it today.

paul-szczepanek-arm committed:
cast no longer needed

paul-szczepanek-arm committed:
deprecate services that are not useless

cmonr commented:
@adbridge Think this is on you to rereview the final(?) changes

cmonr commented:
@adbridge or @ARMmbed/mbed-docs When y'all get a chance today.

paul-szczepanek-arm committed:
reset advertising sets on reset

paul-szczepanek-arm committed:
legacy handle starts off present in _existing_sets

pan- commented on features/FEATURE_BLE/source/generic/GenericGap.cpp:
```diff
@@ -1345,8 +1345,15 @@ ble_error_t GenericGap::startAdvertising(const GapAdvertisingParams &params)
     _scan_timeout.detach();
+
+    _existing_sets.clear();
+    _active_sets.clear();
```
Can we stop all active extended advertising sets ?

pan- commented on features/FEATURE_BLE/source/generic/GenericGap.cpp:
```diff
@@ -1345,8 +1345,15 @@ ble_error_t GenericGap::startAdvertising(const GapAdvertisingParams &params)
+
+    _existing_sets.clear();
+    _active_sets.clear();
+    _active_periodic_sets.clear();
```
Can we stop all active periodic advertising sets ?

pan- commented on features/FEATURE_BLE/source/generic/GenericGap.cpp:
```diff
@@ -1345,8 +1345,15 @@ ble_error_t GenericGap::startAdvertising(const GapAdvertisingParams &params)
+    _existing_sets.clear();
+    _active_sets.clear();
+    _active_periodic_sets.clear();
+    _active_sets.set(LEGACY_ADVERTISING_HANDLE);
```
wrong set, it should be `_existing_sets` and it must be re-created after `clear_advertising_sets`.

paul-szczepanek-arm committed:
stop advertising before reseting

pan- committed:
BLE: Enable cordio extended modules if available.

paul-szczepanek-arm committed:
check feature support before calling extended functnions

pan- commented:
https://github.com/ARMmbed/mbed-os/pull/8738/commits/ac47736259bd4ae4746bb6f949e432a1681c3616 and 
https://github.com/ARMmbed/mbed-os/pull/8738/commits/7f3314576058f50d636631fd40fca0d2fb14021d Enable BT5 extensions on the Cordio Stack if they are available. 

@donatieng @paul-szczepanek-arm Please review.

paul-szczepanek-arm committed:
always set existing set for legacy handle

pan- committed:
BLE: Fix extended advertising check.

adbridge commented:
Latest updates look ok from the maintainers perspective

donatieng reviewed (approved):
Looks good @pan- - this will fully re-init the GAP module but I think that works as this stage as no other `dm` API should have been called in the meantime

pan- committed:
BLE: Fix advertising set issues on legacy controllers.

pan- committed:
BLE: Fix set advertising paylaod dispatch.

pan- committed:
BLE: Report legacy advertising with new API.

paul-szczepanek-arm committed:
fix is feature enabled condition and made in consistent

pan- committed:
BLE: report legacy connection event in new event handler.

paul-szczepanek-arm committed:
set active set to false when stoping advertising

pan- commented:
https://github.com/ARMmbed/mbed-os/pull/8738/commits/4d986bf376cf8d69ef12299bdc84b57665758b01 and 
https://github.com/ARMmbed/mbed-os/pull/8738/commits/a14f143c850dddf77393ce02d08f1b9f503144ce reviewed.

paul-szczepanek-arm committed:
fix return code for implemented func

pan- committed:
BLE: remove unwanted trace.

pan- committed:
BLE: report correct event type for legacy advertising report.

cmonr commented on features/FEATURE_BLE/source/generic/GenericGap.cpp:
```diff
@@ -1551,9 +1551,31 @@ void GenericGap::on_advertising_report(const pal::GapAdvertisingReportEvent &e)
             _eventHandler->onAdvertisingReport(
                 AdvertisingReportEvent(
-                    advertising_event_t(/*advertising.type*/ 0), // TODO
+                    advertising_event_t(event_type), // TODO
```
TODO?

pan- replied:
I fixed the TODO but forgot to remove the TODO. Brilliant. Let me remove that comment.

pan- committed:
BLE: Fix address name in NRF targets.

cmonr commented:
Remainder of commits look fine, aside for the lone `TODO`.

If it's small fix, please correct it. Otherwise, I think the only thing left is a scrub by docs.

pan- committed:
BLE: removed resolved TODO.

pan- commented:
@cmonr Fixed.

Commit:
Edit Gap.h

Edit file for capitalization and active voice.

Commit:
Edit Bounded.h

Edit file, mostly for consistent capitalization and active voice.

AnotherButler commented on features/FEATURE_BLE/ble/common/Duration.h:
```diff
@@ -0,0 +1,578 @@
+ * Represent an integral compile time value that can be used in Duration.
+ *
+ * @tparam T Type of the integral value.
+ * @tparam V The integer value representing a never ending duration.
```
Global note throughout the file: Please make sure "Duration" is capitalized when referring to the class but not otherwise.

AnotherButler commented on features/FEATURE_BLE/ble/common/Duration.h:
```diff
@@ -0,0 +1,578 @@
+    /**
+     * Construct a Duration from an integer value.
+     *
+     * @param v The value of the duration in TN units.
```
Query: What are TN units?

pan- replied:
That is a typo, it should be `TB` or `TIME_BASE`.

AnotherButler commented on features/FEATURE_BLE/ble/common/Duration.h:
```diff
@@ -0,0 +1,578 @@
+    /**
+     * Construct a Duration from another Duration.
+     *
+     * @note The operation fail at compile time of there is a loss of precision.
```
Query: What does this mean? The operation fails at compile time if there's a loss of precision? Or something else?

pan- replied:
Type: `of` -> `if`

AnotherButler commented on features/FEATURE_BLE/ble/common/Duration.h:
```diff
@@ -0,0 +1,578 @@
+    }
+
+    /**
+     * Return the duration in milliseconds.
```
Global note throughout the file: Sometimes we use "milliseconds", and sometimes we abbreviate. Please standardize.

AnotherButler commented on features/FEATURE_BLE/ble/common/Duration.h:
```diff
@@ -0,0 +1,578 @@
+};
+
+/**
+ * Type that represents micro seconds.
```
Global note throughout the file: Sometimes we use "us", and sometimes we spell it out. Please standardize. (Please do not include a space in "microseconds" if we choose to spell it out.)

AnotherButler commented on features/FEATURE_BLE/ble/common/Duration.h:
```diff
@@ -0,0 +1,578 @@
+ * @tparam RangeIn The range of duration.
+ * @tparam FIn The Forever value of duration.
+ * @param duration The duration to convert.
+ * @return The converted duration. It is rounded up if precision is loss.
```
Query: Should this be "... if precision is lost."?

pan- replied:
Yes .

Commit:
Edit AdvertisingDataBuilder.h

Edit file, mostly to fix typos and standardize hyphen use.

Commit:
Update Duration.h in consistent casing

Commit:
Edit AdvertisitngDataSimpleBuilder.h

Edit file, mostly for consistent tense and typos.

donatieng commented:
@AnotherButler tried to do my best with durations (although it's extremely confusing :))

AnotherButler commented on features/FEATURE_BLE/ble/gap/AdvertisingDataTypes.h:
```diff
@@ -0,0 +1,539 @@
+        UNKNOWN = 0,
+
+        /**
+         * Generic Phone.
```
Most of these descriptions repeat the name. Is there a way to provide more information?

pan- replied:
The specification doesn't offer more precision for these values. We document the entities to make sure they get rendered in Doxygen but the name is already self descriptive.

Commit:
Edit AdvertisingParameters.h

Edit file, mostly for active voice and U.S. spelling.

Commit:
Edit ConnectionParameters.h

Edit file, mostly for agreement.

Commit:
Edit Events.h

Edit file for person, capitalization and to fix typos.

Commit:
Edit Gap.h

Edit most of file, mostly for U.S. spelling and typos.

Commit:
Edit iBeacon.h

Add commas between two independent clauses joined by a coordinating conjunction.

Commit:
Edit URIBeaconCOnfigService.h

Edit file for active voice and comma use.

Commit:
Edit ScanParameters.h

Edit file, mostly to fix typos.

Commit:
Edit Types.h

Edit file.

AnotherButler reviewed (commented):
I left a few queries and requests for changes.

cmonr commented:
CI started!

AnotherButler commented:
I can deal with my queries being addressed in a patch release, but please be sure they are addressed. I don't want them to fall through the cracks.

cmonr commented:
@ARMmbed/mbed-os-maintainers Please be sure that @AnotherButler comments get ack'd in some way before merging this PR.

0xc0170 commented:
> @ARMmbed/mbed-os-maintainers Please be sure that @AnotherButler comments get ack'd in some way before merging this PR.

The new PR would be the best to address them (as this is already in the CI).

0xc0170 commented:
Exporters failed (the rest of CI pipeline looks good), investigating

0xc0170 commented:
Restarted exporter (IAR failed due to multiple writes to a file, will be investigated).

0xc0170 commented:
Reviewed again, found the issue , two Gap.cpp files (IAR can't handle files if you run with compile multiple files - I experienced this before).

pan- committed:
BLE: Move source/Gap.cpp -> source/LegacyGap.cpp

pan- commented:
@0xc0170 I pushed a fix.

0xc0170 commented:
New Ci job started

0xc0170 commented:
Only exporters failed,

There's colorama import failure @OPpuolitaival Please review this job

```
12:04:21 [WIZWIKI_W7500:iar] + aws s3 cp xxxx/xxx.log s3://xx-os-ci/xxxx-ci/xxxx/8738/9/iar/PASS/WIZWIKI_W7500/exporter_build_log_WIZWIKI_W7500_iar.log
12:04:21 [WIZWIKI_W7500:iar] Traceback (most recent call last):
12:04:21 [WIZWIKI_W7500:iar]   File "C:/mbed_tools/Python27/Scripts/aws", line 19, in <module>
12:04:21 [WIZWIKI_W7500:iar]     import awscli.clidriver
12:04:21 [WIZWIKI_W7500:iar]   File "C:\mbed_tools\python27\lib\site-packages\awscli\clidriver.py", line 27, in <module>
12:04:21 [WIZWIKI_W7500:iar]     from awscli.formatter import get_formatter
12:04:21 [WIZWIKI_W7500:iar]   File "C:\mbed_tools\python27\lib\site-packages\awscli\formatter.py", line 20, in <module>
12:04:21 [WIZWIKI_W7500:iar]     from awscli.table import MultiTable, Styler, ColorizedStyler
12:04:21 [WIZWIKI_W7500:iar]   File "C:\mbed_tools\python27\lib\site-packages\awscli\table.py", line 18, in <module>
12:04:21 [WIZWIKI_W7500:iar]     import colorama
12:04:21 [WIZWIKI_W7500:iar] ImportError: No module named colorama
```

0xc0170 commented:
Exporters restarted

OPpuolitaival commented:
@0xc0170 that is CI internal problem, not related to this change

cmonr commented:
Restarting exporter CI
